### PR TITLE
kernel-firmware: use copy-firmware.sh

### PR DESCRIPTION
**Pull request by MilhouseVH**
Since [this commit](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/commit/intel?id=9cfefbd7fbdaa5ae769e3061c463f8345d146fb7), the symbolic links for latest Intel DSP firmware are no longer included in the repository, but will instead be created by running `copy-firmware.sh` (which processes the `WHENCE` file).

This PR avoids the following issue:
```
find: 'intel/dsp_fw_bxtn.bin': No such file or directory
find: 'intel/dsp_fw_cnl.bin': No such file or directory
find: 'intel/dsp_fw_glk.bin': No such file or directory
find: 'intel/dsp_fw_kbl.bin': No such file or directory
find: 'intel/dsp_fw_release.bin': No such file or directory
```

The `copy-firmware.sh` script does a good job, however it is dependent on the `WHENCE` file being correct. For this reason we need to add the `WHENCE` entries for the Broadcom nvram configuration files that are added as a patch (which is why the call to `copy-firmware.sh` takes place in `post_patch()` and not `post_unpack()`). I've left the `WHENCE` change as a separate patch in case it needs updating in future.

When comparing the contents of the latest `linux-firmware` (see #4327) against a directory created by `copy-firmware.sh` I didn't find any issues ie. no missing files that were of any significance -  LICENCE files, NOTICE.txt, authentication signatures and other non-firmware files are not copied by the script, which might save a smidgen of space.

### CODE REVIEW

**Comment by MilhouseVH** [comment]
This could be considered for back porting.

**Comment by HiassofT** [comment]
As commented on Slack the brcmfmac oatch and as a consequence also the WHENCE patch in this PR should be dropped and the brcmfmac nvram files should be moved to another patckage that can then be included by the projects that need/want it.

This allows us to have a clean, patch free kernel-firmware package without having to worry about patch rebasing on bumps

**Comment by MilhouseVH** [comment]
That sounds good. Can we merge this first, and handle that in a follow up PR?

**Comment by HiassofT** [comment]
Is this PR that urgent? I'd rather avoid adding yet another patch that then needs to be removed in a subsequent PR. Remember: every patch added is bad :-)

**Comment by MilhouseVH** [comment]
I'd view this PR as a bug fix. If we decide to bump `linux-firmware` in `libreelec-9.2` then we'll have the same bug there. I'd like to get this tested and merged sooner than later so that we have the option of backporting it (with #4327).

**Comment by HiassofT** [comment]
Does this fix an urgent bug like some user reported breakage or a build breakage? If not there's no need to hurry and better to it properly. Then we won't have to backport 2 PRs.

**Comment by MilhouseVH** [comment]
OK. I don't have any plans to expand this PR beyond what it currently does. If you'd like to push commits to this PR you're most welcome.

**Comment by HiassofT** [comment]
Sorry, don't have time to dig into this ATM.

**Comment by HiassofT** [comment]
I now had a closer look at the brcmfmac firmware loader and the solution is quite easy: drop patches/kernel-firmware-02-add-brcmfmac-43xxx-configs.patch and also the WHENCE patch.

The nvram files are board specific and included in the kernel firmware repo. The fallback to some non-board-specific nvram file won't work in general so there's not much point in adding these.

See also this kernel commit: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=eae8e50669e15002b195177212a6e25afbe7cf4d

**Comment by MilhouseVH** [comment]
That isn't how I understand the original problem, to be honest.

This is the original [forum post](https://forum.libreelec.tv/thread/11249-how-to-change-files-on-read-only-file-system/?postID=104334#post104334) that led us to add the "generic" nvram files as the user has this nvram-related error on an x86_64 system:
```
[   12.074660] brcmfmac: brcmf_fw_map_chip_to_name: using brcm/brcmfmac43455-sdio.bin for chip 0x004345(17221) rev 0x000006
[   12.075734] usbcore: registered new interface driver brcmfmac
[   12.114266] brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac43455-sdio.txt failed with error -2
```

Granted this is with 4.19.4 which pre-dates the [board-specific commit](https://github.com/torvalds/linux/commit/eae8e50669e15002b195177212a6e25afbe7cf4d) (added 5.0-rc1) however based on the wording of the commit if we don't include `brcm/brcmfmac43455-sdio.txt` _somehow_ (either from this package, or a different package) then non-RPi systems such as x86_64 will continue to fail to find a usable `brcmfmac43455-sdio` nvram config, and be unable to use WiFi.

The forum user subsequently confirms that the addition of `brcm/brcmfmac43455-sdio.txt` allows WiFi to work on his system, so if we remove the patch(es) then we're going to break non-RPi systems.

At least, this is all my understanding of the kernel commit - yes, there's a generic nvram fallback but that will only work if a generic nvram file exists (which it doesn't, by default, hence the patch in this package).

**Comment by HiassofT** [comment]
No, this isn't quite how this works. The reason why no brcmfmac nvram files were in the kernel firmware repo back then is that there's no generic nvram file that will work with all boards (as the file is board specific).

This was subsequently fixed in the kernel and nvram files were added to the kernel firmware repo.

Back then we had no possibility to add firmware files via .config/firmware, which has been added, too.

So what we basically did is add a generic file, that worked for this one specific user, and it probably will fail for everyone else.

The correct solution now is that if there's no firmware available from the kernel repo the user will have to extract his nvram file from the windows driver and put it into .config/firmware - or nag the vendor to submit it to the firmware repo.

TL;DR: there's a reason why there are not fallback nvram files in the repo, and we shouldn't add them either

**Comment by MilhouseVH** [comment]
OK, thanks for bearing with me. :smile:

Patches dropped.

**Comment by HiassofT** [comment]
Thanks! Can you squash the WHENCE patch commit and the brcmfmac drop commit? Then it should be good to go IMO

**Comment by MilhouseVH** [comment]
Squashed.

**Comment by MilhouseVH** [comment]
I've added an additional commit so that the original problem of non-existent firmware files does not go unnoticed in future. Now, the package will fail:
```
find: 'intel/dsp_fw_bxtn.bin': No such file or directory
find: 'intel/dsp_fw_cnl.bin': No such file or directory
find: 'intel/dsp_fw_glk.bin': No such file or directory
find: 'intel/dsp_fw_kbl.bin': No such file or directory
find: 'intel/dsp_fw_release.bin': No such file or directory
ERROR: Firmware pattern does not exist: intel/dsp_fw_{bxtn,cnl,glk,kbl,release}.bin
FAILURE: scripts/build kernel-firmware during makeinstall_target (package.mk)
*********** FAILED COMMAND ***********
${SCRIPTS}/build "${1}" "${PARENT_PKG}"
**************************************
```

**Review by geeebse — COMMENTED** [review]
brcm/brcmfmac43455-sdio

**Review by geeebse — COMMENTED** [review]
;)
